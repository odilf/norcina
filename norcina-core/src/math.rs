//! Mostly generic combinatoric functions.

/// The factorial function: n * (n - 1) * (n - 2) * ... * 1
pub const fn fac(n: usize) -> usize {
    if n == 0 { 1 } else { n * fac(n - 1) }
}

/// `n` choose `m`. That is, n! / (m! * (n - m)!)
pub const fn choose(n: usize, m: usize) -> usize {
    fac(n) / fac(m) / fac(n - m)

    // let mut output = 1;

    // // n * (n - 1) * (n - 2) * ... * (n - m + 1)
    // let mut i = n;
    // while i > n - m {
    //     output *= i;
    //     i -= 1;
    // }

    // // m * (m - 1) * (m - 2) * ... * 1
    // let mut i = m;
    // while i > 1 {
    //     output *= i;
    //     i -= 1;
    // }

    // output
}
