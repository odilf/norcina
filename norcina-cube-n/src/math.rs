use std::{
    fmt::{self, Write as _},
    mem::transmute,
};

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Direction {
    Positive = 0,
    Negative = 1,
}

impl Direction {
    #[inline]
    pub const fn from_bool(value: bool) -> Self {
        // SAFETY: The valid bit-patterns for `bool` are 0b0 and 0b1, which are precisely the valid values for Direction.
        unsafe { transmute(value) }
    }

    #[inline]
    const fn bool(self) -> bool {
        // SAFETY: The valid bit-patterns for `bool` are 0b0 and 0b1, which are precisely the valid values for Direction.
        unsafe { transmute(self) }
    }

    #[inline]
    pub const fn u8(self) -> u8 {
        self as u8
    }

    /// Maps 0 -> Positive, everything else -> Negative.
    #[inline]
    pub const fn from_u8_any(value: u8) -> Direction {
        Direction::from_bool(value != 0)
    }

    /// # Safety
    ///
    /// `value` needs to be either 0 or 1.
    #[inline]
    pub const unsafe fn from_u8_unchecked(value: u8) -> Direction {
        debug_assert!(value < 2);
        unsafe { transmute(value) }
    }

    /// Maps 0 -> Positive, 1 -> Negative.
    ///
    /// Panics value is other than 0 or 1.
    #[inline]
    pub const fn from_u8(value: u8) -> Direction {
        assert!(value < 2);
        unsafe { Self::from_u8_unchecked(value) }
    }

    #[inline]
    pub const fn flip(self) -> Self {
        unsafe { Self::from_u8_unchecked(self.u8() ^ 0b1) }
    }
}

impl std::ops::BitXor for Direction {
    type Output = Direction;
    fn bitxor(self, rhs: Self) -> Self::Output {
        // TODO: Check is unnecessary, pretty sure we know answer is valid statically.
        Self::from_u8(self.u8() ^ rhs.u8())
    }
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Axis {
    X = 0,
    Y = 1,
    Z = 2,
}

impl Axis {
    #[inline]
    pub const fn u8(self) -> u8 {
        self as u8
    }

    #[inline]
    pub const fn from_u8(value: u8) -> Axis {
        assert!(value < 3);
        // SAFETY: `Axis` is a u8 and the value can only be 0, 1 or 2.
        unsafe { Axis::from_u8_unchecked(value) }
    }

    #[inline]
    pub const fn from_u8_mod3(value: u8) -> Axis {
        // SAFETY: We have taken modulo 3, so the possible values are 0, 1 or 2.
        unsafe { Axis::from_u8_unchecked(value % 3) }
    }

    /// # Safety
    ///
    /// Value needs to be either 0, 1 or 2.
    pub const unsafe fn from_u8_unchecked(axis_index: u8) -> Axis {
        debug_assert!(axis_index < 3);
        unsafe { transmute(axis_index) }
    }

    #[inline]
    pub const fn next(self) -> Self {
        // TODO: Is there a more efficient way?
        // 00 -> 01
        // 01 -> 10
        // 10 -> 00
        Axis::from_u8_mod3(self.u8() + 1)
    }

    #[inline]
    pub const fn prev(self) -> Axis {
        // TODO: Is there a more efficient way?
        // 00 -> 10
        // 10 -> 01
        // 01 -> 00
        Axis::from_u8_mod3(self.u8() + 2)
    }

    /// An axis that is neither `a` or `b`.
    ///
    /// It is kind of "cross product"-y. See also [`Face::cross`]
    pub const fn other(a: Axis, b: Axis) -> Axis {
        // TODO: Is this good?
        //    a - (b - a)
        // => 2a - b
        // => 2a + 3 - b (to prevent underflow)
        Axis::from_u8_mod3((2 * a.u8()) + 3 - b.u8())
    }
}

// TODO: Implement everything with transmuting.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Face {
    /// Right
    R = 0,
    /// Up
    U = 1,
    /// Front
    F = 2,
    /// Left
    L = 4,
    /// Down
    D = 5,
    /// Back
    B = 6,
}

impl Face {
    pub const fn u8(self) -> u8 {
        self as u8
    }

    pub const fn new(axis: Axis, direction: Direction) -> Face {
        Face::from_u8(axis.u8() + (direction.u8() << 2))
    }

    pub const fn from_u8(index: u8) -> Face {
        assert!(index != 3 && index < 7);
        match index {
            0 => Face::R,
            1 => Face::U,
            2 => Face::F,
            4 => Face::L,
            5 => Face::D,
            6 => Face::B,
            _ => panic!("Invalid face index"),
        }
    }

    #[inline]
    pub const fn axis(self) -> Axis {
        Axis::from_u8(self.u8() & 0b011)
    }

    #[inline]
    pub const fn direction(self) -> Direction {
        Direction::from_bool(self.u8() & 0b100 != 0)
    }

    #[inline]
    pub const fn opposite(self) -> Face {
        Face::from_u8(self.u8() ^ 0b100)
    }

    /// "cross product" of faces. I.e., takes two perpendicular faces and returns another perpendicular face
    /// that follows the right-hand rule.
    ///
    /// # Panics
    ///
    /// If `self` and `rhs` are not perpendicular.
    pub fn cross(self, rhs: Face) -> Face {
        assert_ne!(self.axis(), rhs.axis());
        let axis = Axis::other(self.axis(), rhs.axis());
        let direction = Direction::from_bool(
            !(self.axis().next() == rhs.axis()) ^ self.direction().bool() ^ rhs.direction().bool(),
        );
        Face::new(axis, direction)
    }

    pub fn iter() -> impl Iterator<Item = Self> {
        [Face::R, Face::U, Face::F, Face::L, Face::D, Face::B].into_iter()
    }

    fn char(self) -> char {
        match self {
            Self::R => 'R',
            Self::U => 'U',
            Self::F => 'F',
            Self::L => 'L',
            Self::D => 'D',
            Self::B => 'B',
        }
    }
}

impl fmt::Display for Face {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_char(self.char())
    }
}

#[cfg(feature = "quickcheck")]
mod quickcheck_impl {
    use super::*;
    use quickcheck::{Arbitrary, Gen};

    impl Arbitrary for Direction {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose([Direction::Positive, Direction::Negative].as_slice())
                .unwrap()
        }
    }
    impl Arbitrary for Axis {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose([Axis::X, Axis::Y, Axis::Z].as_slice()).unwrap()
        }
    }

    impl Arbitrary for Face {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose([Face::R, Face::U, Face::F, Face::L, Face::D, Face::B].as_slice())
                .unwrap()
        }
    }
}

#[cfg(all(test, feature = "quickcheck"))]
mod tests {
    use super::*;
    use quickcheck::{TestResult, quickcheck};
    quickcheck! {
        fn other_axis_is_always_different_axis(a: Axis, b: Axis) -> TestResult {
            let other = Axis::other(a, b);
            if a == b {
                return TestResult::discard();
            }

            TestResult::from_bool(other != a && other != b)
        }

        // fn cross_product_follows_right_hand_rule(a: Face, direction: bool) -> bool {
        //     let b = if direction { a.next() } else { a.next().next() };
        //     a.cross(b).direction() == Direction::from_bool(direction)
        //     // let expected = if a.axis().next() == b.axis() { Direction::Positive } else { Direction::Negative };
        //     // a.cross(b).direction() == expected
        // }

        fn fn_axis_u8_is_always_between_0_and_2(x: Axis) -> bool {
            x.u8() < 3
        }

        fn fn_face_u8_has_always_direction_in_bit_3_and_axis_in_1_and_2(axis: Axis, direction: Direction) -> bool {
            let face = Face::new(axis, direction);
            (((face.u8() & 0b100) >> 2) == direction.u8())
                && (face.u8() & 0b11 == axis.u8())
        }
    }
}
