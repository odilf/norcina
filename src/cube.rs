use std::fmt::{self, Write as _};

use owo_colors::{OwoColorize, Rgb};

use crate::{
    corner::{self, Corner, CornerPosition},
    edge::{self, Edge, EdgePosition},
    math::Face,
    mov::Move,
};

#[derive(Clone, Copy, PartialEq, Eq)]
pub struct Cube {
    corners: [Corner; 8],
    edges: [Edge; 12],
}

impl Cube {
    pub const SOLVED: Self = Cube {
        corners: [
            Corner::solved(0),
            Corner::solved(1),
            Corner::solved(2),
            Corner::solved(3),
            Corner::solved(4),
            Corner::solved(5),
            Corner::solved(6),
            Corner::solved(7),
        ],
        edges: [
            Edge::solved(0),
            Edge::solved(1),
            Edge::solved(2),
            Edge::solved(3),
            Edge::solved(4),
            Edge::solved(5),
            Edge::solved(6),
            Edge::solved(7),
            Edge::solved(8),
            Edge::solved(9),
            Edge::solved(10),
            Edge::solved(11),
        ],
    };

    fn sticker_at(self, face: Face, up: Face, col: i32, row: i32) -> Sticker {
        // Center sticker
        if col == 1 && row == 1 {
            return face;
        }

        let side = up.cross(face);

        if (col + row) % 2 == 0 {
            let faces = [
                face,
                if row == 0 { up } else { up.opposite() },
                if col == 0 { side.opposite() } else { side },
            ];

            let position = CornerPosition::from_faces(faces);
            let piece = position.pick(self.corners);
            corner::sticker(piece, position, face)
        } else {
            let other_face = match (row, col) {
                (0, 1) => up,
                (1, 0) => side.opposite(),
                (1, 2) => side,
                (2, 1) => up.opposite(),
                _ => unreachable!(),
            };

            let position = EdgePosition::from_faces([face, other_face]);
            let piece = position.pick(&self.edges);
            edge::sticker(piece, position, face)
        }
    }

    pub fn mov(self, mov: Move) -> Self {
        Self {
            corners: corner::move_pieces(self.corners, mov),
            edges: edge::move_pieces(self.edges, mov),
        }
    }
}

pub type Sticker = Face;
pub type ColorScheme = fn(Face) -> Rgb;

const DEFAULT_COLOR_SCHEME: ColorScheme = |face| match face {
    Face::R => Rgb(217, 39, 39),
    Face::U => Rgb(250, 250, 250),
    Face::F => Rgb(109, 242, 116),
    Face::L => Rgb(255, 153, 12),
    Face::D => Rgb(255, 224, 0),
    Face::B => Rgb(79, 123, 212),
};

impl fmt::Debug for Cube {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        struct CornersDebug([Corner; 8]);
        impl fmt::Debug for CornersDebug {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                let mut dl = f.debug_list();
                for (i, corner) in self.0.into_iter().enumerate() {
                    let position = CornerPosition::from_index(i as u8);
                    dl.entry(&format_args!("{corner} is at {position}"));
                }

                dl.finish()
            }
        }

        struct EdgesDebug([Edge; 12]);
        impl fmt::Debug for EdgesDebug {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                let mut dl = f.debug_list();
                for (i, edge) in self.0.into_iter().enumerate() {
                    let position = EdgePosition::from_index(i as u8);
                    dl.entry(&format_args!("{edge} is at {position}"));
                }

                dl.finish()
            }
        }

        f.debug_struct("Cube")
            .field("corners", &CornersDebug(self.corners))
            .field("edges", &EdgesDebug(self.edges))
            .finish()
    }
}

impl fmt::Display for Cube {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let write = |f: &mut fmt::Formatter<'_>, sticker| {
            write!(f, "{}", "██".color((DEFAULT_COLOR_SCHEME)(sticker)))
        };

        let pad = |f: &mut fmt::Formatter<'_>| f.write_str("      ");

        for u_row in 0..3 {
            pad(f)?;
            for u_col in 0..3 {
                write(f, self.sticker_at(Face::U, Face::B, u_col, u_row))?;
            }
            f.write_char('\n')?;
        }

        for lfr_row in 0..3 {
            for l_col in 0..3 {
                write(f, self.sticker_at(Face::L, Face::U, l_col, lfr_row))?;
            }
            for f_col in 0..3 {
                write(f, self.sticker_at(Face::F, Face::U, f_col, lfr_row))?;
            }
            for r_col in 0..3 {
                write(f, self.sticker_at(Face::R, Face::U, r_col, lfr_row))?;
            }

            f.write_char('\n')?;
        }

        for d_row in 0..3 {
            pad(f)?;
            for d_col in 0..3 {
                write(f, self.sticker_at(Face::D, Face::F, d_col, d_row))?;
            }
            f.write_char('\n')?;
        }
        for b_row in 0..3 {
            pad(f)?;
            for b_col in 0..3 {
                write(f, self.sticker_at(Face::B, Face::D, b_col, b_row))?;
            }
            f.write_char('\n')?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use quickcheck::{quickcheck, Arbitrary, Gen};

    use super::*;

    impl Arbitrary for Cube {
        fn arbitrary(g: &mut Gen) -> Self {
            Self {
                corners: [Corner::arbitrary(g); 8],
                edges: [Edge::arbitrary(g); 12],
            }
        }
    }
}
