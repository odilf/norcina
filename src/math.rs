// TODO: Do this with bit manipulations and transmute
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Direction {
    Positive = 0,
    Negative = 1,
}

impl Direction {
    #[inline]
    pub const fn from_bool(value: bool) -> Self {
        // TODO: Do this with bit manipulations and transmute
        if value {
            Direction::Negative
        } else {
            Direction::Positive
        }
    }

    #[inline]
    const fn bool(self) -> bool {
        // TODO: Do this with bit manipulations and transmute
        match self {
            Direction::Positive => false,
            Direction::Negative => true,
        }
    }

    #[inline]
    pub const fn u8(self) -> u8 {
        self as u8
    }

    #[inline]
    pub const fn is_positive(self) -> bool {
        // TODO: Do this with bit manipulations and transmute
        matches!(self, Direction::Positive)
    }

    #[inline]
    pub const fn is_negative(self) -> bool {
        // TODO: Do this with bit manipulations and transmute
        matches!(self, Direction::Negative)
    }

    #[inline]
    pub const fn flip(self) -> Self {
        // TODO: Do this with bit manipulations and transmute
        match self {
            Direction::Positive => Direction::Negative,
            Direction::Negative => Direction::Positive,
        }
    }

    pub const fn from_u8(axis_offset: u8) -> Direction {
        // TODO: Do this with bit manipulations and transmute
        Self::from_bool(axis_offset != 0)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Axis {
    X = 0,
    Y = 1,
    Z = 2,
}

impl Axis {
    #[inline]
    pub const fn next(self) -> Self {
        Axis::from_u8((self as u8 + 1) % 3)
    }

    #[inline]
    pub const fn prev(self) -> Axis {
        Axis::from_u8((self as u8 + 2) % 3)
    }

    #[inline]
    pub const fn from_u8(axis_index: u8) -> Axis {
        match axis_index {
            0 => Axis::X,
            1 => Axis::Y,
            2 => Axis::Z,
            _ => panic!("Axis index was greater than 3"),
        }
    }

    /// An axis that is neither `a` or `b`.
    ///
    /// It is kind of "cross product"-y. See also [`Face::cross`]
    pub const fn other(a: Axis, b: Axis) -> Axis {
        // TODO: Is this good?
        //    a - (b - a)
        // => 2a - b
        // => 2a + 3 - b (to prevent underflow)
        Axis::from_u8(((2 * a.u8()) + 3 - b.u8()) % 3)
    }

    pub const fn u8(self) -> u8 {
        self as u8
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Face {
    /// Right
    R = 0,
    /// Up
    U = 1,
    /// Front
    F = 2,
    /// Left
    L = 4,
    /// Down
    D = 5,
    /// Back
    B = 6,
}

impl Face {
    pub const fn u8(self) -> u8 {
        self as u8
    }

    pub const fn new(axis: Axis, direction: Direction) -> Face {
        Face::from_u8(axis.u8() + (direction.u8() << 2))
        // match (axis, direction) {
        //     (Axis::X, Direction::Positive) => Face::R,
        //     (Axis::Y, Direction::Positive) => Face::U,
        //     (Axis::Z, Direction::Positive) => Face::F,
        //     (Axis::X, Direction::Negative) => Face::L,
        //     (Axis::Y, Direction::Negative) => Face::D,
        //     (Axis::Z, Direction::Negative) => Face::B,
        // }
    }

    pub const fn from_u8(index: u8) -> Face {
        match index {
            0 => Face::R,
            1 => Face::U,
            2 => Face::F,
            4 => Face::L,
            5 => Face::D,
            6 => Face::B,
            _ => panic!("Invalid face index"),
        }
    }

    #[inline]
    pub const fn axis(self) -> Axis {
        Axis::from_u8(self.u8() & 0b011)
    }

    #[inline]
    pub const fn direction(self) -> Direction {
        Direction::from_bool(self.u8() & 0b100 != 0)
    }

    #[inline]
    pub const fn opposite(self) -> Face {
        Face::from_u8(self.u8() ^ 0b100)
    }

    /// "cross product" of faces. I.e., takes two perpendicular faces and returns another perpendicular face
    /// that follows the right-hand rule.
    ///
    /// # Panics
    ///
    /// If `self` and `rhs` are not perpendicular.
    pub fn cross(self, rhs: Face) -> Face {
        assert_ne!(self.axis(), rhs.axis());
        let axis = Axis::other(self.axis(), rhs.axis());
        let direction = Direction::from_bool(
            !(self.axis().next() == rhs.axis()) ^ self.direction().bool() ^ rhs.direction().bool(),
        );
        Face::new(axis, direction)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use quickcheck::{Arbitrary, Gen, TestResult, quickcheck};

    impl Arbitrary for Direction {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose([Direction::Positive, Direction::Negative].as_slice())
                .unwrap()
        }
    }
    impl Arbitrary for Axis {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose([Axis::X, Axis::Y, Axis::Z].as_slice()).unwrap()
        }
    }

    impl Arbitrary for Face {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose([Face::R, Face::U, Face::F, Face::L, Face::D, Face::B].as_slice())
                .unwrap()
        }
    }

    quickcheck! {
        fn other_axis_is_always_different_axis(a: Axis, b: Axis) -> TestResult {
            let other = Axis::other(a, b);
            if a == b {
                return TestResult::discard();
            }

            TestResult::from_bool(other != a && other != b)
        }

        // fn cross_product_follows_right_hand_rule(a: Face, direction: bool) -> bool {
        //     let b = if direction { a.next() } else { a.next().next() };
        //     a.cross(b).direction() == Direction::from_bool(direction)
        //     // let expected = if a.axis().next() == b.axis() { Direction::Positive } else { Direction::Negative };
        //     // a.cross(b).direction() == expected
        // }

        fn fn_axis_u8_is_always_between_0_and_2(x: Axis) -> bool {
            x.u8() < 3
        }

        fn fn_face_u8_has_always_direction_in_bit_3_and_axis_in_1_and_2(axis: Axis, direction: Direction) -> bool {
            let face = Face::new(axis, direction);
            (((face.u8() & 0b100) >> 2) == direction.u8())
                && (face.u8() & 0b11 == axis.u8())
        }
    }
}
