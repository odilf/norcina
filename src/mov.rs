use std::{fmt, ops};

use crate::math::{Axis, Direction, Face};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Amount {
    Single = 1,
    Double = 2,
    Reverse = 3,
}

impl Amount {
    #[inline]
    pub const fn u8(self) -> u8 {
        self as u8
    }

    #[inline]
    pub const fn from_u8(x: u8) -> Self {
        match x {
            1 => Self::Single,
            2 => Self::Double,
            3 => Self::Reverse,
            _ => panic!("Invalid amount."),
        }
    }

    fn iter() -> impl Iterator<Item = Amount> {
        [Amount::Single, Amount::Double, Amount::Reverse].into_iter()
    }
}

impl ops::Mul<Direction> for Amount {
    type Output = Amount;
    fn mul(self, rhs: Direction) -> Self::Output {
        // TODO: Is it possible to remove this branch?
        match rhs {
            Direction::Positive => self,
            Direction::Negative => Self::from_u8(4 - self.u8()),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Move {
    /// Packed field: `---aafff`
    data: u8,
}

impl Move {
    pub const fn new(face: Face, amount: Amount) -> Move {
        Self {
            data: face.u8() + (amount.u8() << 3),
        }
    }

    pub const fn face(self) -> Face {
        Face::from_u8(self.data & 0b111)
    }

    pub const fn amount(self) -> Amount {
        Amount::from_u8(self.data >> 3)
    }

    pub const fn axis(self) -> Axis {
        self.face().axis()
    }

    /// Enumerates all possible moves.
    pub fn iter() -> impl Iterator<Item = Self> {
        Face::iter().flat_map(|face| Amount::iter().map(move |amount| Move::new(face, amount)))
    }
}

impl fmt::Display for Move {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let amount_str = match self.amount() {
            Amount::Single => " ",
            Amount::Double => "2",
            Amount::Reverse => "'",
        };

        write!(f, "{}{}", self.face(), amount_str)
    }
}

#[macro_export]
macro_rules! alg {
    (@ $mov:tt) => { $mov };

    ($($mov:tt)*) => {{
        use $crate::mov::moves::*;
        [$(alg!(@ $mov)),*]
    }};

}

pub mod moves {
    macro_rules! generate_moves {
        (@amount 1) => { Amount::Single };
        (@amount 2) => { Amount::Double };
        (@amount 3) => { Amount::Reverse };

        (@face R) => { Face::R };
        (@face U) => { Face::U };
        (@face F) => { Face::F };
        (@face L) => { Face::L };
        (@face D) => { Face::D };
        (@face B) => { Face::B };

        ([$($face:tt $amount:tt $name:ident),*]) => {
            $(
                pub const $name: Move = Move::new(
                    generate_moves!(@face $face),
                    generate_moves!(@amount $amount)
                );
            )*
        };
    }

    use super::Amount;
    use super::Face;
    use super::Move;

    generate_moves!([
        R 1 R, R 2 R2, R 3 RP,
        U 1 U, U 2 U2, U 3 UP,
        F 1 F, F 2 F2, F 3 FP,
        L 1 L, L 2 L2, L 3 LP,
        D 1 D, D 2 D2, D 3 DP,
        B 1 B, B 2 B2, B 3 BP
    ]);
}

pub mod algs {
    use super::Move;

    pub const SEXY: [Move; 4] = alg!(R U RP UP);
    pub const SLEDGEHAMMER: [Move; 4] = alg!(RP F R FP);

    pub const T: [Move; 14] = alg!(R U RP UP RP F R2 UP RP UP R U RP FP);
    pub const J: [Move; 13] = alg!(R U RP F R U RP UP RP FP R2 UP RP);
    pub const U: [Move; 11] = U_A;
    pub const U_A: [Move; 11] = alg!(R2 UP RP UP R U R U R UP R);
    pub const U_B: [Move; 11] = alg!(RP U RP UP RP UP RP U R U R2);

    pub const CHECKER: [Move; 6] = alg!(R2 L2 U2 D2 F2 B2);

    // TODO: Concat or extend algs
    // pub const J_AUF: [Move; 14] = [J, alg!(UP)].concat();
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cube::Cube;
    use quickcheck::{Arbitrary, Gen, quickcheck};

    impl Arbitrary for Amount {
        fn arbitrary(g: &mut Gen) -> Self {
            *g.choose(&[Amount::Single, Amount::Double, Amount::Reverse])
                .unwrap()
        }
    }

    impl Arbitrary for Move {
        fn arbitrary(g: &mut Gen) -> Self {
            let face = Face::arbitrary(g);
            let amount = Amount::arbitrary(g);
            Move::new(face, amount)
        }
    }

    quickcheck! {
        fn fn_move_constructor_and_accessors_maintain_values(face: Face, amount: Amount) -> bool {
            let mov = Move::new(face, amount);
            mov.face() == face && mov.amount() == amount
        }

        fn fn_double_double_identity(cube: Cube, face: Face) -> bool {
            let mov = Move::new(face, Amount::Double);
            cube.mov([mov, mov]) == cube
        }

        fn fn_rrp_identity(cube: Cube) -> bool {
            cube.mov(alg!(R RP)) == cube
        }

        fn fn_move_reverse_identity(cube: Cube, face: Face) -> bool {
            let m1 = Move::new(face, Amount::Single);
            let m2 = Move::new(face, Amount::Reverse);
            cube.mov([m1, m2]) == cube
        }

        fn fn_double_t_identity(cube: Cube) -> bool {
            cube.mov(alg!(R RP)) == cube
        }

        fn fn_single_double_equals_reverse(cube: Cube) -> bool {
            cube.mov(alg!(R R2)) == cube.mov(alg!(RP))
        }
    }

    #[test]
    fn instance_all_basic_moves() {
        for mov in Move::iter() {
            insta::assert_debug_snapshot!(Cube::SOLVED.mov_single(mov))
        }
    }

    #[test]
    fn ua_ub_cancel() {
        assert!(Cube::SOLVED.mov(algs::U_A).mov(algs::U_B).is_solved())
    }
}
