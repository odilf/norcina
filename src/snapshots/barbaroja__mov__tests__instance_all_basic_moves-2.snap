---
source: src/mov.rs
expression: "Cube::SOLVED.mov_single(mov)"
---
Cube {
    corners: [
        RDB (0) is at RUF,
        LUF (0) is at LUF,
        RUB (0) is at RDF,
        LDF (0) is at LDF,
        RDF (0) is at RUB,
        LUB (0) is at LUB,
        RUF (0) is at RDB,
        LDB (0) is at LDB,
    ],
    edges: [
        UF (✓) is at UF,
        DF (✓) is at DF,
        UB (✓) is at UB,
        DB (✓) is at DB,
        BR (✓) is at FR,
        FR (✓) is at BR,
        FL (✓) is at FL,
        BL (✓) is at BL,
        RD (✓) is at RU,
        LU (✓) is at LU,
        RU (✓) is at RD,
        LD (✓) is at LD,
    ],
}
